src/renderer.js

```diff
@@ -2,47 +2,64 @@
 
 const CDP = require('chrome-remote-interface');
 
-function extractHead(url) {
-  return new Promise((resolve, reject) => {
-    CDP.New().then((tab) => {
-      return CDP({tab: tab});
-    }).then((client) => {
-      try {
-        // Extract DevTools domains.
-        const {Page, Runtime, Network} = client;
-
-        // Enable events on domains we are interested in.
-        Promise.all([
-          Page.enable(),
-          Runtime.enable(),
-          Network.enable(),
-        ]).then(() => {
-          return Promise.all([
-            Network.clearBrowserCache(),
-            Network.setCacheDisabled({cacheDisabled: true}),
-            Network.setBypassServiceWorker({bypass: true}),
-          ]);
-        }).then(() => {
-          Page.navigate({url: url});
-        });
-
-        // Load and dump DOM of head element.
-        Page.loadEventFired(() => {
-          setTimeout(async() => {
-            let result = await Runtime.evaluate({expression: 'document.head.outerHTML'});
-            CDP.Close({id: client.tab.id});
-            resolve(result.result.value);
-          }, 1500);
-        });
-      } catch (err) {
-        console.error(err);
-        CDP.Close({id: client.tab.id});
-        reject(err);
+function render(url) {
+  return new Promise(async(resolve, reject) => {
+    const tab = await CDP.New();
+    const client = await CDP({tab: tab});
+
+    const {Page, Runtime, Network, Emulation} = client;
+
+    await Promise.all([
+      Page.enable(),
+      Runtime.enable(),
+      Network.enable(),
+      Network.clearBrowserCache(),
+      Network.setCacheDisabled({cacheDisabled: true}),
+      Network.setBypassServiceWorker({bypass: true}),
+    ]);
+
+    Page.navigate({url: url});
+
+    // Check that all outstanding network requests have finished loading.
+    const outstandingRequests = new Map();
+    Network.requestWillBeSent((event) => {
+      outstandingRequests.set(event.requestId, event);
+    });
+
+    Network.loadingFinished((event) => {
+      outstandingRequests.delete(event.requestId);
+    });
+
+    Network.loadingFailed((event) => {
+      outstandingRequests.delete(event.requestId);
+    });
+
+    // Ensure the page load event has fired.
+    let pageLoadEventFired = false;
+    Page.loadEventFired(() => {
+      pageLoadEventFired = true;
+    });
+
+    // Set a virtual time budget of 10 seconds. This 10 second timer is paused while there are
+    // any active network requests. This allows for a maximum of 10 seconds in script/rendering
+    // time. Once the page is idle, the virtual time budget expires immediately.
```

Comment:
> Once the page is idle, the virtual time budget expires immediately

For my understanding, is it just how the virtual time thing works that it expires immediately and runs the callback when the page is "idle"?  How is "idle" defined? Couldn't find that in a cursory look at the API docs.